refactor combined_editor and add combined_editor_array

matortheeternal opened the pull request:
This pull request refactors the implementation of the `combined_editor` scripting function and introduces a simplified variant `combined_editor_array` which accepts an array of fields and a single separator as an argument.

## Motivation

The existing `combined_editor` function had grown into a large, monolithic block of code that was difficult to maintain, reason about, or extend.  Although the initial goal was just extending it to support supplying an array of fields as an argument, refactoring it became a necessary intermediate step.

The lack of a way to dynamically pass collections of fields forced template authors to enumerate `field1`, `field2`, etc., which made more complex templates (such as Sagas and Planeswalkers) cumbersome to maintain.

## Changes
**Refactored `combined_editor`**
- Extracted the function’s major logical sections into well-named static helpers:
  - `readFieldArguments`
  - `readSeparatorArguments`
  - `removePrefixSuffix`
  - `splitTheValue`
  - `updateValuesIfInputNewer`
  - `recombineParts`
  - `handleEmpty`

**Refactored `SCRIPT_FUNCTION_DEPENDENCIES(combined_editor)`**
- Split into helper functions:
  - `readDepFieldArguments`
  - `resolveTargetField`
  - `addFieldDependencies`

**Added new Function: `combined_editor_array`**
- Adds support for passing an array of fields and a single separator string, e.g.:
```
combined_editor_array(fields: [card.level_1_text, card.level_2_text, card.level_3_text], separator:"<line>\n</line>")
```
- Allows dynamic generation of combined text from variable-length field sets.
- Supports both direct arrays and variables (e.g., `fields:ar`).

## Benefits

- Improved the maintainability and readability of the `combined_editor` function.  This will make it easier for future changes to be built on demand.
- Enables more flexible and compact scripts that can dynamically compose `combined_editor` fields.
- This work can set the tone for future refactoring of other scripting functions with similar structure.

## Validation

Validation is currently in progress.  I have done initial validation with the major fields on the mainframe template: Name, Power/Toughness, Type/Subtype, flavor text, and card text.  @CajunAvenger is currently doing more broad validation as we speak.

Parity with the following behavior should be verified:
- Prefix/suffix stripping
- Separator insertion 
  - ✔️ power/toughness
  - ✔️ supertype/subtype
- Empty-field handling (hide_when_empty, soft_before_empty) 
  - ✔️ power/toughness
  - ✔️ supertype/subtype

The new combined_editor_array functionality should be tested in real-world template use-cases.

## tl;dr

In short: this PR cleans up one of the larger and more complex functions in the scripting subsystem and adds a requested array-based API for template authors.

The PR will remain a draft until full verification by Cajun of parity and expected functionality has been performed.

matortheeternal committed:
refactor combined_editor and add combined_editor_array

CajunAvenger commented:
* Save/load
  ✔️Properly loading field value
  ✔️Properly saving field value without prefix/suffix interference
  ✔️Fields are updating real time
  ✔️Field dependencies working
  ✔️Tagged strings working
* Separator insertion
  ✔️ power/toughness
  ✔️ supertype/subtype
  ✔️`<line></line>` lines
  ✔️collector number
 * Empty-field handling (hide_when_empty, soft_before_empty)
  ✔️ power/toughness
  ✔️ supertype/subtype
* Verified against a number of MSEM/Rev set variants on Sagas, Classes, Levelers, and Planeswalkers
* combined_editor_array
  ❌ collection of fields is not being recognized, even as a direct parameter. field values are being coerced to normal strings per mator
  ❌ getting 'parameter not set: fields_param', might be from error above
  ❌  error message still refers to it as 'x_editor'

CajunAvenger commented:
* combined_editor_array
  ✔️ collection of fields is recognized, even as a variable parameter.
  ✔️ no longer getting 'parameter not set: fields_param'
  ✔️ error message corrected
  ✔️ other behaviors still clean

matortheeternal committed:
added min-height tag

matortheeternal committed:
added min_heights param to combined_editor_array

matortheeternal committed:
idk it's working helly aea

matortheeternal committed:
fix bug with not using min_height tag wrapping on first section

matortheeternal committed:
fix last block margin handling

matortheeternal committed:
officially resolved all edge cases for top & bottom margin handling

matortheeternal committed:
update docs for combined_editor_array

Co-authored-by: cajun <12363371+CajunAvenger@users.noreply.github.com>

G-e-n-e-v-e-n-s-i-S commented:
Getting some weird behavior with this test template:

[test.mse-style.zip](https://github.com/user-attachments/files/23378099/test.mse-style.zip)

If I click anywhere in the region that I have highlighted in red here:

<img width="377" height="525" alt="combined_editor" src="https://github.com/user-attachments/assets/b8708e5c-4f91-4ec8-b977-146bef1c581e" />

it selects the first textbox, when it should select the second

matortheeternal committed:
fixed scope handling in the dependency engine

previously, jumps could create bugs in the dependency engine where a scope would be incorrectly closed, resulting in variables prior to conditional expressions being nullified.  this fix resolves the problem by creating scopes around jumps.

CajunAvenger commented:
> @CajunAvenger if I understand correctly, the issue is the chapter icon positioning on top/bottom boxes is not operating as expected, but it works correctly in middle boxes? this is something that should be fixed on my side (?)

the Top of Block 0 and the Top of the Textbox are not necessarily the same value (same for bottom of the last Block and bottom of the Textbox)

When Min-Heights does the alignment, it aligns with invisible top/bottom of the Block instead of the visible top/bottom of the textbox.

For those images, the reminder text isn't in combined_editor, and alignment is set to middle. But the first and last chapter aren't visually middle aligned, only chapters in between them are. The saga dividers there are drawn where the combined editor outlines are.

I would think having the first and last alignment use the textbox top and textbox bottom rather than block top and block bottom would resolve this?
